lib/DXIL/DxilOperations.cpp

```diff
@@ -2876,6 +2876,128 @@ bool OP::IsDxilOpFeedback(OpCode C) {
+      // If the memory type is AllMemory, masking inapplicable flags is allowed.
+      if (memoryTypeFlags != (unsigned)DXIL::MemoryTypeFlag::AllMemory &&
+          (memoryTypeFlags & (unsigned)DXIL::MemoryTypeFlag::GroupSharedMemory))
+        return true;
+    }
+  }
+    LLVM_FALLTHROUGH;
+  case OpCode::BarrierByMemoryHandle:
+  case OpCode::BarrierByNodeRecordHandle: {
+    // BarrierByMemoryType, BarrierByMemoryHandle, and BarrierByNodeRecordHandle
+    // all have semanticFlags as the second operand.
+    DxilInst_BarrierByMemoryType barrier(const_cast<CallInst *>(CI));
+    if (isa<ConstantInt>(barrier.get_SemanticFlags())) {
+      unsigned semanticFlags = barrier.get_SemanticFlags_val();
+      if (semanticFlags & ((unsigned)DXIL::BarrierSemanticFlag::GroupScope |
+                           (unsigned)DXIL::BarrierSemanticFlag::GroupSync))
+        return true;
+    }
+    return false;
+  }
+  default:
+    return false;
+  }
+}
+
+bool OP::BarrierRequiresNode(const llvm::CallInst *CI) {
+  OpCode opcode = OP::GetDxilOpFuncCallInst(CI);
+  switch (opcode) {
+  case OpCode::BarrierByNodeRecordHandle:
+    return true;
+  case OpCode::BarrierByMemoryType: {
+    DxilInst_BarrierByMemoryType barrier(const_cast<CallInst *>(CI));
+    if (isa<ConstantInt>(barrier.get_MemoryTypeFlags())) {
+      unsigned memoryTypeFlags = barrier.get_MemoryTypeFlags_val();
+      // If the memory type is AllMemory, masking inapplicable flags is allowed.
+      return memoryTypeFlags != (unsigned)DXIL::MemoryTypeFlag::AllMemory &&
+             (memoryTypeFlags &
+              ((unsigned)DXIL::MemoryTypeFlag::NodeInputMemory |
+               (unsigned)DXIL::MemoryTypeFlag::NodeOutputMemory));
+    }
+    return false;
+  }
+  default:
+    return false;
+  }
+}
+
+DXIL::BarrierMode OP::TranslateToBarrierMode(const llvm::CallInst *CI) {
+  OpCode opcode = OP::GetDxilOpFuncCallInst(CI);
+  switch (opcode) {
+  case OpCode::Barrier: {
+    DxilInst_Barrier barrier(const_cast<CallInst *>(CI));
+    if (isa<ConstantInt>(barrier.get_barrierMode())) {
+      unsigned mode = barrier.get_barrierMode_val();
+      return static_cast<DXIL::BarrierMode>(mode);
+    }
+    return DXIL::BarrierMode::Invalid;
+  }
+  case OpCode::BarrierByMemoryType: {
+    unsigned memoryTypeFlags = 0;
+    unsigned semanticFlags = 0;
+    DxilInst_BarrierByMemoryType barrier(const_cast<CallInst *>(CI));
+    if (isa<ConstantInt>(barrier.get_MemoryTypeFlags())) {
+      memoryTypeFlags = barrier.get_MemoryTypeFlags_val();
+    }
+    if (isa<ConstantInt>(barrier.get_SemanticFlags())) {
+      semanticFlags = barrier.get_SemanticFlags_val();
+    }
+    // If the memory type is AllMemory, masking inapplicable flags is allowed.
+    if (memoryTypeFlags != (unsigned)DXIL::MemoryTypeFlag::AllMemory) {
+      if (memoryTypeFlags & ((unsigned)DXIL::MemoryTypeFlag::NodeInputMemory |
+                             (unsigned)DXIL::MemoryTypeFlag::NodeOutputMemory))
```

Comment:
Nit: Having this exception in multiple places makes me a little nervous. It might be safer to have a utility function that does this for you.